Add multipart machines

BarryNorfolk opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
Add a component/entity system for handling multipart machines in a more generalized way.

## Why / Balance
Multipart machines are cool, and it'd be nice to see more than one in the game.

## Technical details
Tried a simple and probably naive approach to finding parts and make it agnostic to whatever part people build first.
At the moment, if you attempt to build everything BUT the fuel chamber then the UI will show that there are no matching entities, which is a bit frustrating as a new player (I can imagine).

So here it's just a simple look in four directions and attempt to find any anchored entities that match our required offset, rotation, and expected components. After that it's up to the PA system to determine if the emitters are also in the right place.

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->
Tested this orientations and they all worked/fired, mirrored orientations do not work however. So placing the control box on the other side doesn't pick up the entities.
![image](https://github.com/user-attachments/assets/92ddb17c-4f52-4d7d-bd2a-d4d2a92d302d)

Also, examining an entity with the multipart machine will try and show you where to place the specific machine components based on the sprites specified in the yaml.

Default rotations will likely be the norm:
![image](https://github.com/user-attachments/assets/94d29eb2-c866-49eb-a041-54d4d606f3f4)

Until a user places at lease one entity and causes a rescan, which means we store the orientation of the machine and can display updated ghosts:
![image](https://github.com/user-attachments/assets/e7f0353a-8b16-486e-99bd-da35adb803ef)

Also added the ability to have optional parts. So in this example, I made the end cap of the PA an optional part.
![image](https://github.com/user-attachments/assets/fc30b492-f056-4a1c-af24-566f5336411e)

Here's a video showing the event based updating of the PA systems.
[MultipartMachine - ItWorks.webm](https://github.com/user-attachments/assets/3d9bbe6b-c629-4914-b872-c5405a2e07c1)

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
:cl:
- add: New component and system for creating more multipart machines!
- tweak: Converted the Particle Accelerator to use new multipart machine system.
- add: Examining a Particle Accelerator now shows ghosts of what and where to place the exact parts.

ArtisticRoomba commented:
Extremely good stuff, this will allow much more creative future machines. It's actually a component in a lot of proposals I plan to write in the future.

But what it applies to right now is being able to make large machines like station anchors and station gravity generators constructible, which is good for SS14's dynamic environment.

ArtisticRoomba commented:
This will also close (or at least work towards) issues https://github.com/space-wizards/space-station-14/issues/34505, https://github.com/space-wizards/space-station-14/issues/18451.

metalgearsloth commented:
OH god it's so good.

Killerqu00 reviewed (approved):
the system behind this looks very good, me likey

Pronana commented:
![image](https://github.com/user-attachments/assets/301d6d3b-6df2-4446-8d54-5238c5924bd8)
![image](https://github.com/user-attachments/assets/b3b15afc-7089-491c-9e10-3617cd3dbc58)

Found an issue(?)
The new system doesn't seem to care about the rotation of the PA computer which is nice, but it also has some weird behavior with predicting a valid multipart machine. Nothing game-breaking just a bit weird.

As posted in the screenshot, regardless of orientation it's detecting the starboard machine part as a port machine part and declaring it valid, regardless of how the computer is rotated and regardless of the rotation of the fuel chamber next to the computer.

Also the computer currently wants to be on the right side of the Particle Accelerator which breaks sprites as the wires connect to the left side of the Particle Accelerator. 

I like the idea of the computer not caring as much about how its rotated because it always faces forward and can be confusing but it feels like it's not working properly.

Pronana commented:
![image](https://github.com/user-attachments/assets/b0e336d6-64e5-4cfb-a9e0-873b0c9218a5)
Actually it appears to only check one direction regardless of the rotation of the PA console. This should be valid but it isn't.

Pronana commented:
Unless I'm dumb or missed something, I left a review that should fix the issues I found. If you have questions feel free to ask.

TiniestShark commented:
Oh the ghost showing how it's supposed to line up is just _perfect_. That's gonna be insanely helpful.

BarryNorfolk commented:
> ![image](https://private-user-images.githubusercontent.com/66055347/425690812-b0e336d6-64e5-4cfb-a9e0-873b0c9218a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI2MjU0MTcsIm5iZiI6MTc0MjYyNTExNywicGF0aCI6Ii82NjA1NTM0Ny80MjU2OTA4MTItYjBlMzM2ZDYtNjRlNS00Y2ZiLWE5ZTAtODczYjBjOTIxOGE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIyVDA2MzE1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2NWM0YzU2YWFiY2ZlYzYxNmMzMDhlNDlkNDI0OGNlN2M3MjQwZTcwODhjNzFmM2RkOGRiMGFkOTcyMWQ4Y2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.nHvKQ28OjEdXgfhrBT_p_daRbEHmAQpTnHfozMehu2Q) Actually it appears to only check one direction regardless of the rotation of the PA console. This should be valid but it isn't.
> 
> Edit to avoid duplicating comments, you can also spam click inspect to duplicate the ghost sprties
> 
> Edit 2: After further testing and reading the code it appears that the issue is that the machine never actually removes the rotation data it stores if it finds even one valid machine part. This means if it finds a part, even if that part is out of place it will be locked to that specific rotation basically forever or until it is deconstructed. Except you can't deconstruct the computer I believe.

You are likely right and I missed it when I quickly added the networked orientation for the ghost sprites. Thanks for looking into it and I appreciate the testing :)

Pronana reviewed (changes requested):
Just some yaml changes needed, otherwise looks fine to me.

BarryNorfolk commented:
> Just some yaml changes needed, otherwise looks fine to me.

Sure, is this so that it just faces the other direction, yeah?

Pronana commented:
> Sure, is this so that it just faces the other direction, yeah?

Yep so it lines up with the original sprites. In its current iteration the wires in the power box/fuel chamber (I mix up the names forgive me) don't connect to the computer, this flips the side the PA is built on so that those wires connect and so it also matches the sprite in the PA computer's UI.

Pronana reviewed (approved):
Looks good to me. Might be worth refactoring the PA logic in the future but that's out of scope for this PR.

Fildrance commented on Content.Client/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,83 @@
+public sealed class MultipartMachineSystem : EntitySystem
+{
+    [ValidatePrototypeId<EntityPrototype>]
+    private const string GhostPrototype = "MultipartMachineGhost";
```
Shouldnt it be ProtoId<EntityPrototype>?

BarryNorfolk replied:
I’ve no idea, I can take a look at others but this mostly just stolen from the TEG ghosts which could be wrong or old.

BarryNorfolk replied:
Should be fixed

BarryNorfolk commented:
I rebased like a nerd, oops.

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.Parts.cs:
```diff
@@ -16,6 +17,25 @@ private void InitializePartSystem()
     }
 
+    public void ValidateEmitter(string name,
+        ParticleAcceleratorEmitterType type,
```
i understand that this is an attempt to faithfully recreate previous code after refactor, but do we really need 3 separate parts for emitters and check their position here? maybe we can dump it and just check it all in scanning for construction complete with some more generic solution? (i saw check for orientation in yaml so thats kinda solved, right?)

BarryNorfolk replied:
We can use the graph/node I added in one of your previous comments which should solve this :)

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.Parts.cs:
```diff
@@ -24,137 +44,43 @@ public void RescanParts(EntityUid uid, EntityUid? user = null, ParticleAccelerat
-                part = entity;
-                return true;
-            }
+            if (partQuery.TryGetComponent(GetEntity(part.Entity), out var partData))
```
my educated guess is that it could be nice to have 'core' part in every multipart machine and move this code into multipart machine system - so parts could always know thier master just in case.

BarryNorfolk replied:
Added this, so now there's a new MultipartMachinePartComponent that allows entities to do other things and know what their "Master" is.

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -244,30 +256,35 @@ public void UpdateUI(EntityUid uid, ParticleAcceleratorControlBoxComponent? comp
+                comp.SelectedStrength,
+                (int)draw,
+                (int)receive,
+                _multipartMachine.GetPartEntity(machine, "StarboardEmitter") != null,
```
looks like HasPart would be nice addition.

BarryNorfolk replied:
Added

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -244,30 +256,35 @@ public void UpdateUI(EntityUid uid, ParticleAcceleratorControlBoxComponent? comp
+        if (!TryComp<MultipartMachineComponent>(uid, out var machineComp))
+            return;
+
+        var machine = new Entity<MultipartMachineComponent?>(uid, machineComp);
```
just 
```var machine = (uid, machineCompe);```

 should be nuff, right?

BarryNorfolk replied:
Neat, in most places yes it seems!

Fildrance commented on Content.Client/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,82 @@
+        var xformQuery = GetEntityQuery<TransformComponent>();
+        foreach (var part in ent.Comp.Parts.Values)
+        {
+            if (!part.Entity.HasValue)
```
prolly better to invert if, use early return here to reduce code nesting

```
if (part.Entity.HasValue)
    continue;
```

BarryNorfolk replied:
Good catch :)

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.Parts.cs:
```diff
@@ -16,6 +17,25 @@ private void InitializePartSystem()
         SubscribeLocalEvent<ParticleAcceleratorPartComponent, PhysicsBodyTypeChangedEvent>(BodyTypeChanged);
```
maybe its better to leave handling of such events to MultipartMachineSystem? we can expect that most of them would require handling it in some way?

Fildrance replied:
I'd say that those events are kinda what will lead to 'complete' and 'disassembled' events that this system is expected to emit

BarryNorfolk commented:
Still looking to respond to the comments from you @Fildrance, just been a long two weeks.
I've looked into the push mechanism and while I can definitely listen for anchor events on constructionComponents, there's no good concept of a machine being "finished" that would allow me to listen in. All I can really do is listen for `ConstructionChangeEntityEvent` and rescan to see if there's a valid entity (Might be timing based though).

Fildrance commented:
Listening and rescanning is the thing i was expecting this thing to do, to be honest - so other systems can not listen and just sub for 'complete' or 'destroyed / non-functional'. That will greatly simplify connected systems code and logic, i feel.

BarryNorfolk commented on Content.Shared/Machines/Components/MultipartMachine.cs:
```diff
@@ -0,0 +1,75 @@
+    /// Dictionary of unique names to specific parts of this machine
+    /// </summary>
+    [DataField, AutoNetworkedField]
+    public Dictionary<string, MachinePart> Parts = [];
```
Done. Had to move the enum/enum values to a shared folder, else the client gets a bit sad when trying to parse the yaml for the control box. Good to know.

BarryNorfolk commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,232 @@
+
+    /// <summary>
+    /// Performs a rescan of all parts of the machine to confirm they exist and match
+    /// the specified requirments for offset, rotation, and components.
```
Fixed.

BarryNorfolk commented:
Think I've fixed or changed everything you mentioned @Fildrance.
Rebased due to merge conflicts but should be good now. Added a webm video showcasing it but I don't know how to properly embed it....

Fildrance commented on Content.Server/Construction/ConstructionSystem.Graph.cs:
```diff
@@ -453,4 +460,19 @@ public ConstructionChangeEntityEvent(EntityUid newUid, EntityUid oldUid)
+    /// <param name="graph">Construction graph for this entity</param>
+    /// <param name="previousNode">Previous node that was active on the graph</param>
+    /// <param name="currentNode">New node that has become active</param>
+    public sealed class AfterConstructionChangeEntityEvent(string graph, string currentNode, string? previousNode)
```
record struct, no parent, ByRefEventAttribute, use primary constructor. 

Also having so many strings is a bit sad, but no good idea what to do with it - maybe some of them are actually ProtoEntId? eh?

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -40,14 +44,16 @@ public override void Update(float frameTime)
-        DebugTools.Assert(EntityManager.EntityExists(controller.PortEmitter));
-        DebugTools.Assert(EntityManager.EntityExists(controller.ForeEmitter));
-        DebugTools.Assert(EntityManager.EntityExists(controller.StarboardEmitter));
+        DebugTools.Assert(machine.Comp.Assembled);
```
shouldn't just checking for Assembled be enough? why check parts separately?

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -58,20 +64,25 @@ public void Fire(EntityUid uid, TimeSpan curTime, ParticleAcceleratorControlBoxC
         comp.NextFire = curTime + comp.ChargeTime;
 
-        EverythingIsWellToFire(comp);
+        if (!TryComp<MultipartMachineComponent>(uid, out var machineComp))
```
trycomp is repeated before each EverythingIsWellToFire call - just put it inside?

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -82,8 +93,9 @@ public void SwitchOn(EntityUid uid, EntityUid? user = null, ParticleAcceleratorC
+        if (!TryComp<PowerConsumerComponent>(_multipartMachine.GetPartEntity(uid, AcceleratorParts.PowerBox), out var powerConsumer)
+            || powerConsumer.ReceivedPower >=
+            powerConsumer.DrawRate * ParticleAcceleratorControlBoxComponent.RequiredPowerRatio)
             PowerOn(uid, comp);
```
please add curly braces here, as it melts into multiline if statement a bit - hard to read.

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -82,8 +93,9 @@ public void SwitchOn(EntityUid uid, EntityUid? user = null, ParticleAcceleratorC
-        || powerConsumer.ReceivedPower >= powerConsumer.DrawRate * ParticleAcceleratorControlBoxComponent.RequiredPowerRatio)
+        if (!TryComp<PowerConsumerComponent>(_multipartMachine.GetPartEntity(uid, AcceleratorParts.PowerBox), out var powerConsumer)
+            || powerConsumer.ReceivedPower >=
+            powerConsumer.DrawRate * ParticleAcceleratorControlBoxComponent.RequiredPowerRatio)
```
join it with previous line, please

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -223,7 +238,7 @@ private void UpdatePowerDraw(EntityUid uid, ParticleAcceleratorControlBoxCompone
     {
         if (!Resolve(uid, ref comp))
             return;
```
empty line after if block

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -244,30 +259,35 @@ public void UpdateUI(EntityUid uid, ParticleAcceleratorControlBoxComponent? comp
+
         _uiSystem.SetUiState(uid,
             ParticleAcceleratorControlBoxUiKey.Key,
             new ParticleAcceleratorUIState(
```
extract variable pls

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -244,30 +259,35 @@ public void UpdateUI(EntityUid uid, ParticleAcceleratorControlBoxComponent? comp
+                comp.SelectedStrength,
+                (int)draw,
+                (int)receive,
+                _multipartMachine.HasPart(machine, AcceleratorParts.StarboardEmitter),
```
this could be resolved on the client, as MultipartMachineComponent is networked with all of parts dictionary

maybe it should be?... coz this State is a phat thingy. Maybe needs to be consulted with Emo or Sloth :<

BarryNorfolk replied:
Discussed a little on discord and I think it _could_ be done but it makes things a bit odd. 

The UI state messages appear to be processed first, so the components aren't updated by the time we handle that message. To make this work we'd either have to have a split UI update from the server, one after the component has updated which updates any preview and then another after the UI state comes in which updates the on/off or power states. Or move it into a frame update tick, with the same level of desync.

I agree sending this all down the pipe from the server feels a bit crap for the moment. I can give it a try anyway but not sure how it'll turn out.

BarryNorfolk replied:
Hrm, this is a bit annoying in how it's structured and I feel like I would be reworking a large amount of the particle accelerator UI than I'd like :/

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -292,39 +312,25 @@ private void UpdatePartVisualStates(EntityUid uid, ParticleAcceleratorControlBox
-            _appearanceSystem.SetData(controller.ForeEmitter!.Value, ParticleAcceleratorVisuals.VisualState, state);
-        if (controller.StarboardEmitter.HasValue)
-            _appearanceSystem.SetData(controller.StarboardEmitter!.Value, ParticleAcceleratorVisuals.VisualState, state);
+        if (_multipartMachine.TryGetPartEntity(machine, AcceleratorParts.FuelChamber, out var fuelChamber) && fuelChamber.HasValue)
```
isnt it fixed by multipart code? there is some sprite-related stuff there! :o

BarryNorfolk replied:
Don't understand what would be fixed here by the multipart code? This is the PA system deciding how to set its UI state depending on what parts are available.

Fildrance commented on Content.Shared/Machines/Components/MultipartMachine.cs:
```diff
@@ -0,0 +1,99 @@
+}
+
+[RegisterComponent, NetworkedComponent, AutoGenerateComponentState]
+public sealed partial class MultipartMachineComponent : Component
```
file should be named after this component, i think.

Fildrance replied:
this should restrict access by its System - add AccessAttribute(typeof(SharedMultipartMachineSystem))

u need to add SharedMultipartMachineSystem for that, so client and server could interact. The idea is that components HAVE to use public fields for seriazliation and interaction with its own system, but for INCAPSULATION it should restrict interaction with its public fields for other systems (so PA should not have access or read its fields, it should ask MultipartMachineSystem for info, or listen to its events)

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -336,14 +342,40 @@ private void OnComponentShutdown(EntityUid uid, ParticleAcceleratorControlBoxCom
+
         var partQuery = GetEntityQuery<ParticleAcceleratorPartComponent>();
-        foreach (var part in AllParts(uid, comp))
+        foreach (var part in machine.Parts.Values)
```
this system should not iterate over parts :< this code probably should be method for MultipartMachineSystem?

BarryNorfolk replied:
Should have been removed, missed that. The `ParticleAcceleratorPartComponent` is now defunct.

Fildrance commented on Content.Server/ParticleAccelerator/EntitySystems/ParticleAcceleratorSystem.ControlBox.cs:
```diff
@@ -397,9 +429,11 @@ private void OnUIRescanMessage(EntityUid uid, ParticleAcceleratorControlBoxCompo
 
-        UpdateUI(uid, comp);
+        var machine = new Entity<MultipartMachineComponent>(uid, machineComp);
+        _multipartMachine.Rescan(machine, msg.Actor); // Raises an event if state has changed
```
This is not a nice hack :( mainly because i dont see how this should raise event - construction not changed, anchored is not changed - it should not raise event just in case Rescan was called. It is not obvious why is this connected to to OnMachineAssembledChanged, this is a hard-to-understand hack, better link this code directly.

BarryNorfolk replied:
I don't believe this is a hack, as the rescan now will just emit an event if the machine's state has changed (new/removed parts. I feel it's better to let the UI message here ask the for the rescan and let the flow of event handling do the job of updating the UI state. 
Else I have to have the rescan here return a bool but also have an optional to parameter to control whether an event has been raised or not. Perhaps I can make the comment here clearer if that makes sense.

BarryNorfolk replied:
Made the comment a bit clearer that this is a manual rescan raised by the user, and may have no effect because we (hopefully) catch all anchor/unanchor/construction events as you point out.
Anything more than this and I'm getting into UI changes for the particle accelerator which I'm not comfortable with and feel like it's scope creeping.

Fildrance reviewed (changes requested):
Code is nice, but we can do better!

BarryNorfolk commented on Content.Server/Construction/ConstructionSystem.Graph.cs:
```diff
@@ -407,6 +409,11 @@ public bool ChangeNode(EntityUid uid, EntityUid? userUid, string id, bool perfor
+            // If ChangeEntity has ran, then the entity uid has changed and the
+            // new entity should be initialized by this point.
+            var afterChangeEv = new AfterConstructionChangeEntityEvent(construction.Graph, construction.Node, previousNode);
+            RaiseLocalEvent(uid, afterChangeEv);
```
Blind on my part, should be `newUid` since that's the new entity that's been created as part of the construction graph move.

BarryNorfolk commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,364 @@
+        var missingParts = false;
+        for (var i = 0; i < ent.Comp.Parts.Count; ++i)
+        {
+            var part = ent.Comp.Parts.Values.ElementAt(i);
```
I think I encountered some oddness with the things being readonly or at least not modify-able, but this seems to work. Thanks.

BarryNorfolk commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,364 @@
+                    {
+                        // This entity succeeds, store the direction we used to get this one and expect all
+                        // future machine parts to match this direction.
+                        ent.Comp.Rotation = curAngle;
```
That wasn't my intention no, my intention was to have the main entities rotation be independent from the machine entirely so you could have it face any direction. Maybe we want to swap that which would then eliminate the rotation check here completely.

I don't really mind or have any strong opinion on it.

BarryNorfolk replied:
Makes sense tbh, I forgot that the PA has a rotation with the wires.

BarryNorfolk commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,364 @@
+            if (!args.Anchored)
+            {
+                // Some construction is being unanchored, check if its a known part for us
+                foreach (var part in machine.Parts.Values)
```
I think it's better to have the event still trigger so things can properly listen out for it, rather than altering the components IsAssembled state or creating a new event for it.
I like the idea of just changing the unanchored part with a master though, that is an easier time and reduces the cost of doing this.

BarryNorfolk replied:
I clearly wrote that in reponse to something but I have no idea anymore, don't code when tired folks.

BarryNorfolk commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,364 @@
+                continue;
+            }
+
+            part.Entity = GetNetEntity(entity);
```
Sadly I think having the EntityUid inside a class that's USED by a component makes the thing not work.
EntityUid isn't marked as serializable so it falls over when you try and run the thing.

BarryNorfolk commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,364 @@
+                if (args.Graph == part.Graph &&
+                    (args.PreviousNode == part.ExpectedNode || args.CurrentNode == part.ExpectedNode))
+                {
+                    Rescan((uid, machine));
```
We could look at the maximum distance from any part that's specified in the YAML and create a "max range" that we epxect things to be in.
That should enable us to not scan EVERY machine which could be expensive. I'll look into this later.

BarryNorfolk replied:
I think I've fixed this so we don't scan ALL the machines all the time.

BarryNorfolk commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,364 @@
+        ent.Comp.Assembled = !missingParts;
+        if (stateHasChanged)
+        {
+            var ev = new MultipartMachineAssemblyStateChanged(ent, ent.Comp.Assembled, user);
```
Sure, that's a neat idea.

Fildrance commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,286 @@
+
+        SubscribeLocalEvent<MultipartMachineComponent, ComponentStartup>(OnComponentStartup);
+
+        SubscribeLocalEvent<ConstructionComponent, AfterConstructionChangeEntityEvent>(OnConstructionNodeChanged);
```
it'd greatly lower amount calls if we apply MultipartMachinePartComponent on prototype level, and and subscribe to 
```
SubscribeLocalEvent<MultipartMachinePartComponent , AfterConstructionChangeEntityEvent> 
SubscribeLocalEvent<MultipartMachinePartComponent , AnchorStateChangedEvent>(OnConstructionAnchorChanged);
```

Fildrance reviewed (changes requested):
Just a little bit more. Most importantly we would like to avoid subbing for ConstructionComponent component, as it will lead to multiple calls, making server busy with not really useful work : D

BarryNorfolk commented on Content.Shared/Machines/Components/MultipartMachineComponent.cs:
```diff
@@ -1,10 +1,48 @@
+    /// Optional parts are not taken into account.
+    /// </summary>
+    [DataField]
+    public bool IsAssembled = false;
```
Only updated in one place so should be fine! Fixed.

Fildrance commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,286 @@
+
+            // Even optional parts should trigger state updates
+            if (part.Entity != originalPart)
+            {
```
slap early return here
```
if (part.Entity == originalPart)
    continue;
````

Fildrance commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,286 @@
+            RaiseLocalEvent(ent, ref ev);
+        }
+
+        Dirty(ent);
```
better to call dirty only if something changed (its not cheap operation) - happily we clearly have markers for that!

Fildrance reviewed (changes requested):
During testing i found out that right after creation PA comuter is displaying all parts 'in place' and rescanning is not helping with it, unless at least 1 part of machine is placed correctly - maybe it'll be easy to fix? Can u look at it? Maybe just init UI empty...

BarryNorfolk commented:
The fix for the 'in place' for the UI is easy and have done it, plus your other two comments.
The hard part is not handling the unanchoring logic so that the PA doesn't continue to fire after the computer is removed from the machine.

If I clear all the parts from the machine when the unanchor event arrives, which is straightforward, and then raise the state change event then the Controlbox logic attempts to update the visual appearance for the parts. Sadly those are now removed so the individual parts stay in the "incorrect" appearance until the computer is re-added. I can add the entities added/removed to the state change message but then updating all the control box logic to handle the extra chance of an event seems... grim.

I'm at a bit of an impasse there with it :/ Either I have to update the entire control box logic to pipe through the event changes or live with the machine just firing even without the PA..

Fildrance reviewed (approved):
**Left some more comments with proposed changes with code** - so u can just cherry-pick it or copy-paste it. Tested in-game with them - looks like it works fine!
I still would love to see if Sloth have any ideas about our range-searching (i'll poke him), but i personally think that its good to go.

metalgearsloth commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,346 @@
+        // so we have to just check everyone in range and perform a rescan.
+        if (!XformQuery.TryGetComponent(ent.Owner, out var constructXform))
+            return;
+
+        var query = EntityQueryEnumerator<MultipartMachineComponent>();
+        while (query.MoveNext(out var uid, out var machineComp))
+        {
+            var machine = (uid, machineComp);
+            if (!IsMachineInRange(machine, constructXform.LocalPosition))
+                continue; // This part is outside the max range of the machine, ignore
+
+            if (Rescan(machine) && HasPartEntity(machine, ent.Owner))
+                return; // This machine is using this entity so we don't need to go any further
+        }
+    }
```
Use EntityLookup as it will vary between entityquery vs range checks as required.

metalgearsloth commented on Content.Shared/Machines/EntitySystems/SharedMultipartMachineSystem.cs:
```diff
@@ -0,0 +1,122 @@
+        var netEnt = GetNetEntity(entity);
+        return HasPartEntity(machine, netEnt);
+    }
+
+    /// <summary>
+    /// Returns whether a machine has a specifed NetEntity bound to one of its parts.
+    /// </summary>
+    /// <param name="machine">Entity, which might have a multpart machine attached, to use for the query.</param>
+    /// <param name="entity">NetEntity to search for.</param>
+    /// <returns>True if any part has the specified NetEntity, false otherwise.</returns>
+    public bool HasPartEntity(Entity<MultipartMachineComponent?> machine, NetEntity entity)
+    {
+        if (!Resolve(machine, ref machine.Comp))
+            return false;
+
+        foreach (var part in machine.Comp.Parts.Values)
+        {
+            if (part.Entity.HasValue && part.Entity.Value == entity)
+                return true;
+        }
+
+        return false;
+    }
+
```
Any reason you don't just use entityuid for these? NetEntity is really only for networking.

BarryNorfolk replied:
You seem to be able to have EntityUid's inside of a component, but not inside any struct that the component uses. So having a list of machine parts like I do means I just can't seem to use these.

BarryNorfolk replied:
Removed now that we rely more on the EntityUid

metalgearsloth commented on Content.Shared/Machines/Components/MultipartMachineComponent.cs:
```diff
@@ -0,0 +1,103 @@
+    [DataField, AutoNetworkedField]
+    public bool IsAssembled = false;
+
+    /// <summary>
+    /// Flag for whether the client side system is allowed to show
+    /// ghosts of missing machine parts.
+    /// Controlled/Used by the client side.
+    /// </summary>
+    [DataField]
+    public List<EntityUid> Ghosts = [];
```
If only the client uses these then delete datafield as client won't serialize them.

If server uses then needs autonetworkedfield.

metalgearsloth commented on Content.Shared/Machines/Components/MultipartMachineComponent.cs:
```diff
@@ -0,0 +1,103 @@
+    /// </summary>
+    [DataField]
+    public Angle Rotation = Angle.Zero;
+
+    /// <summary>
+    /// Network entity associated with this part.
+    /// Not null when an entity is successfully matched to the part and null otherwise.
+    /// </summary>
+    [DataField]
+    public NetEntity? Entity = null;
+
```
This won't save properly and will explode whenever someone tries this with persistence, you're better off just using EntityUid and having a temporary field for the netentity then flagging this as [NonSerialized]

BarryNorfolk replied:
Do you mean that we should cache it locally ok the client with a temporary netent when the component is dirtied?

TheShuEd commented on Resources/Prototypes/Entities/Structures/Power/Generation/PA/control_box.yml:
```diff
@@ -27,6 +27,62 @@
+        offset: 1, -1
+        rotation: -90
+        sprite:
+          sprite: Structures/Power/Generation/PA/end_cap.rsi
```
how about we pass EntProtoId here instead of a sprite? This would pull layered sprites from the SpriteComponent, as well as the description and part name for inspection. It's quite possible to throw in Dummy entities, with a separate custom description and no logic

Fildrance replied:
Exceptional idea, looks great


https://github.com/user-attachments/assets/e55802ad-c26d-4877-8b76-2a2f7a50c588

Fildrance commented on Content.Server/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,324 @@
+            return;
+
+        if (ent.Comp.Parts.TryGetValue(part, out var value))
+            value.Entity = null;
```
yes, also added making part component have null on master and dirtied it.

Fildrance commented on Content.Shared/Machines/Components/MultipartMachinePartComponent.cs:
```diff
@@ -0,0 +1,15 @@
+/// </summary>
+[RegisterComponent]
+public sealed partial class MultipartMachinePartComponent : Component
+{
```
added dirty and networking

Fildrance commented on Content.Shared/Machines/Components/MultipartMachineGhostComponent.cs:
```diff
@@ -0,0 +1,14 @@
+    /// <summary>
+    /// Machine this particular ghost is linked to.
+    /// </summary>
+    public EntityUid? LinkedMachine = null;
```
Moved comp the thing to client - i am sure that its not wanted on server?...

Fildrance commented on Content.Client/Machines/EntitySystems/MultipartMachineSystem.cs:
```diff
@@ -0,0 +1,92 @@
+    {
+        if (ent.Comp.Ghosts.Count != 0)
+        {
+            // Already showing some part ghosts
```
To be honest i think that it could be beneficial to have multiple ghosts - for example if we will have different multipart machines that could interact in some way, and u want to connect them... its useful!
